tests/unit/s2n_send_test.c

```diff
@@ -124,18 +246,360 @@ int main(int argc, char **argv)
+
+        /* Two full sends with no dynamic record size reset. */
+        EXPECT_EQUAL(conn->active_application_bytes_consumed, large_data.size * 2);
+
+        writes += 1;
+
+        /* Here we reset the clock and move the timeout threshold, forcing a reset. */
+        conn->last_write_elapsed = 0;
+        timeout_threshold_secs = 1;
+        EXPECT_SUCCESS(s2n_connection_set_dynamic_record_threshold(conn, resize_threshold, timeout_threshold_secs));
+
+        s2n_custom_send_fn_called = false;
+        conn->last_write_elapsed = 0;
+        EXPECT_SUCCESS(s2n_send(conn, large_data.data, large_data.size, &blocked)); 
+        EXPECT_TRUE(s2n_custom_send_fn_called);
+        EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
+        EXPECT_NOT_EQUAL(conn->last_write_elapsed, 0);
+        EXPECT_EQUAL(conn->active_application_bytes_consumed, large_data.size);
+    }
+
+    /* s2n_send guards against a partial_send that retries with a smaller buffer than the previous s2n_send */
+    {
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(conn);
+        EXPECT_OK(s2n_connection_set_secrets(conn));
+
+        /* Send 1/3 of the large_data buffer then have control come back to the application. */
+        ssize_t total_bytes_to_send = large_data.size / 3;
+        EXPECT_SUCCESS(s2n_connection_set_ctx(conn, (void*)&total_bytes_to_send));
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_counted_send_fn));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(conn, (void*) conn));
+
+        s2n_blocked_status blocked = 0;
+
+        s2n_custom_send_fn_called = false;
+        /* Use large data so multiple records are written. */
+        ssize_t written = s2n_send(conn, large_data.data, large_data.size, &blocked);
+        EXPECT_TRUE(s2n_custom_send_fn_called);
+        EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_WRITE);
+
+        /* Subsequent writes should only be decreasing. */
+        s2n_custom_send_fn_called = false;
+        EXPECT_FAILURE_WITH_ERRNO(s2n_send(conn, large_data.data, written - large_data.size, &blocked), S2N_ERR_SEND_SIZE);
+        EXPECT_TRUE(s2n_custom_send_fn_called);
+        EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_WRITE);
+    }
+
+    /* s2n_send properly tracks bytes across partial sends */
+    {
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(conn);
+        EXPECT_OK(s2n_connection_set_secrets(conn));
+
+        /* Send 1/3 of the large_data buffer then have control come back to the application. */
+        ssize_t total_bytes_to_send = large_data.size / 3;
+        EXPECT_SUCCESS(s2n_connection_set_ctx(conn, (void*)&total_bytes_to_send));
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_counted_send_fn));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(conn, (void*) conn));
```

Comment:
I need to check `conn->current_user_data_consumed` to get the amount of user data sent in the callback :/. I can't rely on the len in the callback because it's encrypted.